### ISSUE TITLE
Security Issue CVE-2021-3163

### ISSUE THREAD

**Issue by hieroshima** [quilljs/quill#3364]
Hi.

I would like to raise a security issue which is described in CVE-2021-3163. Is there any fix for that or do someone know an ETA when that security issue will be fixed?

Thanks in advance.

**Comment by arktronic** [comment]
Has anyone been able to reproduce this vulnerability? I tried using various setups without luck. The originally referenced issue (#3273) has been deleted. The only way I've been able to see the XSS in action is on the CodePen Playground on the bottom of https://quilljs.com/docs/delta/ but that's because, for the right side of the playground, they're assigning directly to `.innerHTML` without escaping (thus, that particular instance is not itself a QuillJS issue).

**Comment by polyatail** [comment]
I have not been able to reproduce this vulnerability either. The closest I've come is editing the contents of the Quill `<div>` as HTML and forcing something like `<div><img src="foobar.jpg" onclick="alert(1234)">foobar</div>` to exist on the page. Even then, copying and pasting that image within the editor, the pasted image box is stripped of the `onclick` property.

**Comment by alexcodito** [comment]
I have been able to reproduce this by loading the unsanitized payload into Quill. Assuming that no additional server-side validation is performed, in a real-world scenario, it would look something like this (e.g. a notes section of a webpage):

1. Save/persist contents of Quill
2. Edit XHR request, replace body's contents with the payload and replay the XHR
3. Reload the page/editor

I have attempted the same steps with several common vectors such as `<script>alert(1)</script>` and `<a href="javascript:alert(1)></a>` without reproducing the XSS. This tells me that Quill makes an effort to prevent XSS and that this is an edge-case that is not being accounted for.

**Comment by polyatail** [comment]
> 1. Save/persist contents of Quill

So to clarify, this issue affects only cases where un-sanitized contents are loaded into a Quill editor via e.g., an API call to the server that reads the contents from a database. It is not possible to directly input this attack into the editor. Correct?

**Comment by alexcodito** [comment]
> 
> 
> > 1. Save/persist contents of Quill
> 
> So to clarify, this issue affects only cases where un-sanitized contents are loaded into a Quill editor via e.g., an API call to the server that reads the contents from a database. It is not possible to directly input this attack into the editor. Correct?

That's correct as far as my understanding/testing goes. 

I wouldn't normally expect Quill to be responsible for security, but since some level of sanitation is already in place, I believe that a simple solution would be to strip all `on*` event attributes. However, it could cause regression if something else is relying on this.

**Comment by polyatail** [comment]
Perhaps a solution here is to pipe the initial input of the editor through the same mechanism that sanitizes clipboard pasting, since that seems to effectively removing these problematic element attributes.

**Comment by alexcodito** [comment]
> 
> 
> Perhaps a solution here is to pipe the initial input of the editor through the same mechanism that sanitizes clipboard pasting, since that seems to effectively removing these problematic element attributes.

That sounds sensible to me. Will you be taking a look at this? If not, I may do a pull-request in the near future.

**Comment by polyatail** [comment]
> That sounds sensible to me. Will you be taking a look at this? If not, I may do a pull-request in the near future.

Feel free to give it a go and tag me for a review.

**Comment by alexcodito** [comment]
It seems that the Quill component/wrapper that we use in my work environment makes use of the `dangerouslyPasteHTML` function when initializing Quill. Now that I am aware of this, I retract my statement about being able to reproduce the XSS vulnerability. I am now also wondering if the person that submitted the CVE report was also making use of the same function. If this is the case, then for obvious reasons, this cannot be called a vulnerability at all. The docs and the name of the function clearly warn about the implications.

Apologies for the confusion. I hope that the individual who submitted the CVE elaborates further. 

On a side note, `dangerouslyPasteHTML` uses `DOMParser().parseFromString` which is intended to fix invalid markup. It also happens to remove *some* dangerous tags such as `<script></script>`, which is why I assumed that Quill was attempting to sanitize input and that the CVE submission was a vector that escapes it. Although our systems are not affected by this in any case (due to server-side sanitation), I have recommended that we switch from using raw HTML to the native Delta/JSON format.

TL;DR: Unless the CVE submitter can demonstrate that XSS is possible without the `dangerouslyPasteHTML` function, then this is a non-issue.

**Comment by arktronic** [comment]
@alexcodito Thank you for following up on that.

**Comment by HamzaAnis** [comment]
> Has anyone been able to reproduce this vulnerability? I tried using various setups without luck. The originally referenced issue (#3273) has been deleted. The only way I've been able to see the XSS in action is on the CodePen Playground on the bottom of https://quilljs.com/docs/delta/ but that's because, for the right side of the playground, they're assigning directly to `.innerHTML` without escaping (thus, that particular instance is not itself a QuillJS issue).

So is this a valid vulnerability? If not then the a lot of people will be getting false potential security vulnerabilities alerts.

**Comment by samcambridge** [comment]
I noticed that the blog post revealing this "exploit" (https://burninatorsec.blogspot.com/2021/04/cve-2021-3163-xss-slab-quill-js.html) mentioned a html editor. Please correct me if i'm wrong but Quill doesn't provide this functionality out of the box.

The only way I've managed to replicate this is using a https://github.com/benwinding/quill-html-edit-button, which describes itself as a `Module which allows you to quickly view/edit the HTML in the editor`.

Pasting the HTML from the blog post into the HTML editor available in this demo https://benwinding.github.io/quill-html-edit-button/src/demos/javascript/demo.html reproduces the issue. Still, that does not make it a Quill issue.

If anyone has had any success reproducing it without custom modules can they let me know how?

**Comment by pauldraper** [comment]
> So is this a valid vulnerability?

> I have been able to reproduce this by loading the unsanitized payload into Quill.

**This is a valid vulnerability if untrusted content is loaded.**

For example, Alice submits content with XSS attack via a website's API. Bob opens a Quill editor with Alice's content and is attacked.

**Comment by HamzaAnis** [comment]
Is anyone planning for a patch on this security Issue ? 

> > So is this a valid vulnerability?
> 
> > I have been able to reproduce this by loading the unsanitized payload into Quill.
> 
> **Yes, this is a valid vulnerability if untrusted content is loaded.**
> 
> For example, Alice submits content with XSS attack via a website's API. Bob opens a Quill editor with Alice's content and is attacked.

**Comment by jkneepkens** [comment]
> > So is this a valid vulnerability?
> 
> > I have been able to reproduce this by loading the unsanitized payload into Quill.
> 
> **Yes, this is a valid vulnerability if untrusted content is loaded.**
> 
> For example, Alice submits content with XSS attack on a website (say, via its API). Bob opens a Quill editor on that website with Alice's content and is attacked.

But shouldn't this be handled by the backend/API which is serving this XSS content to the Quill editor to remove already this malicious content? Otherwise it would be really nice if it gets fixed in a patch!

**Comment by ndite** [comment]
Some feedback regarding plans to address this would be appreciated. `quill` package being flagged as vulnerable we have to document why this is not an issue if it is not being fixed or replace if it is not maintained.

![image](https://user-images.githubusercontent.com/5113965/129954372-054ca60f-6c23-4d44-bd26-fbcfb522c972.png)

**Comment by einfallstoll** [comment]
I had a look at the source code and there aren't many places which allow XSS using `.innerHTML`. From the [https://burninatorsec.blogspot.com/2021/04/cve-2021-3163-xss-slab-quill-js.html](blog post) my best guess is the reporter just included an XSS payload in the editor div (just like documented in the Quickstart) and then the XSS will execute:

```
<!-- Include Quill stylesheet -->
<link href="dist/quill.snow.css" rel="stylesheet">

<!-- Create the toolbar container -->
<div id="toolbar">
  <button class="ql-bold">Bold</button>
  <button class="ql-italic">Italic</button>
</div>

<!-- Create the editor container -->
<div id="editor">
  <image src=validateNonExistantImage.png onerror=alert(1337)> hey girl hey
</div>

<!-- Include the Quill library -->
<script src="dist/quill.js"></script>

<!-- Initialize Quill editor -->
<script>
  var editor = new Quill('#editor', {
    modules: { toolbar: '#toolbar' },
    theme: 'snow'
  });
</script>
```

If this is the actual XSS, the risk is fairly low, because that's not a vulnerability in Quill but intended behavior in ... a browser. To mitigate this:

- don't include any user-input without HTML encoding it first (which you basically _always_ must do)
- use a Content-Security-Policy to prevent _unsafe-inline_ scripts

To me this risk is acceptable and I can understand why no fix is available or will be available.

**Comment by snoopysecurity** [comment]
Hi, I've contacted the CVE database to revoke this CVE with the evidence provided. I will also get this removed from the Snyk Security Database. https://snyk.io/vuln/SNYK-JS-QUILL-1245047. Hopefully that will reduce the noise

**Comment by AM1988** [comment]
> Hi, I've contacted the CVE database to revoke this CVE with the evidence provided. I will also get this removed from the Snyk Security Database. https://snyk.io/vuln/SNYK-JS-QUILL-1245047. Hopefully that will reduce the noise

Hi. Any updates on this? Seems still treated as a vulnerable package.

**Comment by pauldraper** [comment]
To clarify for all the visitors not understanding if this applies to them:

**If you save and load shared Quill content without server-side filtering, you have XSS.**

For example:

Alice submits content with XSS attack via direct HTTP request.
The website saves this content.
Bob edits/views that content on the website.
When the website opens Quill editor, Bob is attacked.

**Comment by Recurse-blip** [comment]
As I understand this is a non-issue and the CVE will be revoked. Is there any progress on this matter?

**Comment by twocs** [comment]
> Hi. Any updates on this? Seems still treated as a vulnerable package.
and
> As I understand this is a non-issue and the CVE will be revoked. Is there any progress on this matter?

It is still indicated as vulnerable.
Github advisory https://github.com/advisories/GHSA-4943-9vgg-gr5r
The CVE advisory is marked as disputed https://nvd.nist.gov/vuln/detail/CVE-2021-3163
However, the SNYK advisory was revoked https://security.snyk.io/vuln/SNYK-JS-QUILL-1245047

**Comment by EricDitp** [comment]
fix would be appreciated.

**Comment by altechzilla** [comment]
Still getting this as a vulnerability alert. What's the resolution here?

**Comment by pauldraper** [comment]
The resolution is to scrub the data from quill before saving in a shared place.

**Comment by Nyamkhuub** [comment]
Still getting this as a vulnerability alert. How to solve it guys?

**Comment by Sliffcak** [comment]
Still getting this as a vulnerability alert

**Comment by deepesh146** [comment]
still getting error any fix?

**Comment by pauldraper** [comment]
No. You can tell by the way this issue is still open, and the alert doesn't include a fix version

**Comment by deepesh146** [comment]
please suggest any other text editor actually i have purchased a theme and quill is built in ,i m building a enterprise app security is most important for company

**Comment by burninatorsec2** [comment]
Hi, this is how to reproduce the issue - you need to edit the traffic with an interception proxy when storing the comment, in order to put this payload in:` <div><image src=validateNonExistantImage.png onloadstart=alert(1337)> thepayload </div>
`
then once it's stored, any other user to visit the page with that payload will get the XSS to pop. This vulnerabililty is very much still active, I've used it in multiple environments since reporting it back in 2021(here https://github.com/quilljs/quill/issues/3273, and here: https://github.com/quilljs/quill/issues/3558).

**Comment by burninatorsec2** [comment]
> The resolution is to scrub the data from quill before saving in a shared place.

Yes, this is one way to remediate it, in lieu of Quill releasing a security patch.

**Comment by burninatorsec2** [comment]
> Perhaps a solution here is to pipe the initial input of the editor through the same mechanism that sanitizes clipboard pasting, since that seems to be effectively removing these problematic element attributes.

Correct, this same sanitation needs to be done server-side/API side, not just in the UI.

**Comment by gustaff-weldon** [comment]
> > Perhaps a solution here is to pipe the initial input of the editor through the same mechanism that sanitizes clipboard pasting, since that seems to be effectively removing these problematic element attributes.

That sounds reasonable. Any change this can be implemented? Or if that's a public method users can call it on their own?

**Comment by burninatorsec2** [comment]
> > Perhaps a solution here is to pipe the initial input of the editor through the same mechanism that sanitizes clipboard pasting, since that seems to be effectively removing these problematic element attributes.
> 
> That sounds reasonable. Any chance this can be implemented? Or if that's a public method users can call it on their own?

Maybe! I know a lot of text editor wysiwyg widgets have issues with XSS, so they all have different solutions that they've implemented different ways. For example, CKEditor introduced a setting "script mode" (or something like that), to either disallow or allow tags, because allowing script tags would open it up any client using that widget to attacks. Perhaps something similar can be done here.

**Comment by polyatail** [comment]
> Hi, this is how to reproduce the issue - you need to edit the traffic with an interception proxy when storing the comment, in order to put this payload in:`<div><image src=validateNonExistantImage.png onloadstart=alert(1337)> thepayload </div>`
> 
> Then once it's stored, any other user to visit the page with that payload will get the XSS to pop. This vulnerability is very much still active, I've used it in multiple environments I've pentested or done bug bounties for since reporting it back in 2021 (here https://github.com/quilljs/quill/issues/3273, and here: #3558).
> 
> EDIT: spelling and wording

@burninatorsec2 As described, this is strictly *not* an issue with QuillJS. If you are able to get a backend to store un-sanitized input, that is an issue with the backend. Blaming Quill for displaying what it's given is like blaming SQL for interpreting `Robert'); DROP TABLE Students;` such that it [allows an injection attack](https://xkcd.com/327/). You are describing an issue with the way Quill is being used in specific applications, rather than an issue with the software.

Asking for a "script mode" to be implemented is a feature request to unburden developers of having to sanitize input on their own, not a bugfix.

**Comment by burninatorsec2** [comment]
@polyatail The burden is first with QuillJS, since other similar text editors found ways to work around the introduced vulnerability. QuillJS could do the same, or find a way to end the dependence on <div> tags for their functionality. So, I disagree. However, you're right that developers of the client apps that use Quill and the developers of Quill should be sanitizing both on the client side and the server side. This applies for any security issue, whether it's command injection, or XSS, or XXE, and that would make a great workaround in the meantime.

Think about in **layers** of protection - if we can trust a 3rd party like Quill to filter inappropriate input AND have each of the developers who use Quill to add their own sanitization, we stand a much better chance of preventing an attack. But the common denominator is Quill here, so that's where the responsibility lies. Usually vendors jumped on fixing security issues that I've reported, because their product has lots of clients that depend on them, and they realize the effect they would have on the client apps and the people that use it. So they implement a fix.

I like the idea of the SQL Injection metaphor, but it doesn't really work here, because no one is blaming how JavaScript works, just like how no reasonable person would ever blame SQL itself for SQL injections. We all already know JavaScript has a lot of problems, so we've been designing around JavaScript security issues for years. Making a library like Quill that relies on raw HTML tags is not a good idea because of stuff like this. They could encode the tags, for example. There are lots of ways to remediate this, that's just one. 

If you want to make a metaphor about preventing this sort of thing for SQL Injections, how about something like this... we, as 3rd party developers, cannot assume that a client is going to be filtering dangerous SQL commands in their client app, or using a WAF. So we do the due diligence and write code with parameterized SQL queries to prevent introducing risk to the client. That's what I would expect from Quill. I realize that the nature of the design may make it difficult to remediate this, and I'm happy to help brainstorm ways to do it because I have experience as both a software engineer and a pen tester.

**Comment by burninatorsec2** [comment]
Personally, I'm not a fan of the "script mode" feature in CKEditor either, since elective security isn't really security (since it makes it a choice instead of a default - maybe make "script mode off" the default?) But it sounds like there's plenty of concerned users in this thread that would use something like that in Quill if it existed to help mitigate the problem.

**Comment by Saduff** [comment]
> As described, this is strictly not an issue with QuillJS. If you are able to get a backend to store un-sanitized input, that is an issue with the backend.

This absolutely is an issue with Quill. The backend _must_ store "un-sanitized" input, otherwise it would be violating data integrity. XSS is _not_ an input problem, it's an output encoding problem (e.g. it's under the "V5.3 Output Encoding and Injection Prevention" subsection in the OWASP ASVS). As a pentester, I often see developers get confused by this.

For example, let's say I want to share an XSS payload here on GitHub. If GitHub "sanitized" the input before saving it to the database then it wouldn't be able to display the payload (in its original form) later, i.e. data integrity has been violated. "Sanitizing" before saving to the database is _always_ an incorrect defense and often creates problems of its own, not just data integrity, but sometimes even security problems. Data is data. What if instead of my XSS payload being displayed on a HTML page, I use GitHub API to retrieve it in JSON instead, what threat does it pose there? It's an output encoding problem, meaning user input must be encoded for the context in which it's used, whether that's HTML, part of an URL, JSON etc. If user input is properly encoded, there is no XSS vulnerability, despite an XSS payload being saved in the database.

Now, as for Quill, it's a WYSIWYG editor which means HTML cannot simply be encoded or the editor wouldn't work properly (you'd no longer "see what you get"). In this case it comes down to this: should Quill support JavaScript execution or not? I would argue that in almost all cases, WYSIWYG editors should _not_ support JS execution. If it's required, it should be configurable and disabled by default. If JS execution should not be supported, but is possible, then that's a bug and not just any bug, a security vulnerability. In this case, Quill _should_ (actually) sanitize the HTML so that JS execution is not possible, similar to what Angular does with its `[innerHTML]` binding. Angular lets you enter certain HTML tags such as `<h1>`, `<img>`, but good luck trying to get JS execution. An alternative would be for the developer to sanitize the HTML using e.g. DOMPurify before giving the HTML as input to Quill, but that should be Quill's job.

> So this is funny - GitHub properly sanitized the div tag I wrote above, instead of allowing it to be displayed.

The opposite actually, GitHub did _not_ sanitize the div tag, which is the reason it didn't show up as text. That's because some HTML tags are allowed in Markdown. Did you think GitHub violated data integrity by removing it?

<div>--> There's a &lt;div&gt; in the DOM here, inspect this element. Other surrounding tags are &lt;p&gt; tags. <--</div><br>

If GitHub sanitized the div tag above, it would not be in the DOM and would show up as the following text instead:
```
<div>--> There's a &lt;div&gt; in the DOM here, inspect this element. Other surrounding tags are &lt;p&gt; tags. <--</div>
```

Why this is not an issue for GitHub and _not_ an XSS vulnerability is because this is a feature. However, there's no feature for executing JavaScript by any means or taking over display of the entire page using CSS in an element's `style` attribute.

**Comment by burninatorsec2** [comment]
Yes, as far as the GitHub example goes - HTML injection by design != XSS, we're in agreement there.

As for Quill:

> but it seems Quill has already properly taken care of it.

Can you point out the commit or version where Quill has implemented a fix? That would be great news and I would love to be able to tell clients that they can upgrade to a version that isn't vulnerable to XSS.

**Comment by Saduff** [comment]
See here for a PoC which proves Quill is not vulnerable:
https://codepen.io/Saduff/pen/PoejxjN

Alert is shown only once and Quill created DOM looks like this:
![image](https://user-images.githubusercontent.com/11700507/191349781-bbfc13de-64a7-4eed-a131-b1c484937037.png)

Notice how the `<p>` tag has no `style` attribute nor event handlers and the `<img>` tag has no `onerror` event handler.

If you have another PoC which shows Quill is vulnerable then please share it.

**Comment by burninatorsec2** [comment]
> 

Yep, the original payload from the blog still works. Though, the one you sent in that link works too, in the latest Firefox browser. So maybe that isn't a very good test, if neither is working for you?

**Comment by burninatorsec2** [comment]
PS is everyone ok with me taking screenshots of this thread for a DEFCON presentation because this would make an amazing talk about how the responsible vulnerability disclosure process works (and sometimes doesn't work)... if not I can redact your usernames, thanks in advance!

**Comment by Saduff** [comment]
> > 
> 
> Yep, the original payload from the blog still works. Though, the one you sent in that link works too, in the latest Firefox browser. So maybe that isn't a very good test, if neither is working for you?

I'm also using latest Firefox. So when you say it "works" for you, you see **two** 'img' alerts pop up in that codepen?

**Comment by burninatorsec2** [comment]
> > Yep, the original payload from the blog still works. Though, the one you sent in that link works too, in the latest Firefox browser. So maybe that isn't a very good test, if neither is working for you?
> 
> I'm also using latest Firefox. So when you say it "works" for you, you see **two** 'img' alerts pop up in that codepen?

Yes, because I added my own payload as well. But like I said, your code pen probably not a great test because it’s all client side, and the stored XSS payload is allowed by Quill and then later displayed by Quill (and thus the alert is popped).

I’m trying to decide if it’s worth the time and  effort to build a test environment like the web apps I’ve been exploiting it in, with a database and backend, or if it would be more convenient to just come across it again in the wild since it’s been so prevalent lately (and make a redacted video).

**Comment by Saduff** [comment]
> Yes, because I added my own payload as well. But like I said, your code pen probably not a great test because it’s all client side, and the stored XSS payload is allowed by Quill and then later displayed by Quill (and thus the alert is popped).

Quill is all client side, what's your point? There's no stored XSS here, there can only be DOM XSS, which there isn't. You can add as many "payloads" as you want and pop as many alerts as you want and then remove Quill and nothing changes.

> I’m trying to decide if it’s worth the time and effort to build a test environment like the web apps I’ve been exploiting it in, with a database and backend, or if it would be more convenient to just come across it again in the wild since it’s been so prevalent lately (and make a redacted video).

That's completely out of scope and has nothing to do with Quill. The XSS you may have seen where Quill is used is caused by something else and would still be there if Quill is completely removed. Like in my case, the cause was `jQuery.html(<user input>)`.

**Comment by burninatorsec2** [comment]
> That's completely out of scope and has nothing to do with Quill. The XSS you may have seen where Quill is used is caused by something else and would still be there if Quill is completely removed. Like in my case, the cause was `jQuery.html(<user input>)`.

Sorry, but that's not correct. These apps don't have XSS in any other fields except for the Quill text editor field. When they remove Quill it removes this problem.

**Comment by Saduff** [comment]
Sure. Produce a PoC to prove it then.

**Comment by burninatorsec2** [comment]
Yes that’s right, two alert dialog boxes.

**Comment by BarkovskiyMaxim** [comment]
Hi! I have solved this issue on the client side with `iframe` and `sandbox` (typescript): 

```ts

    function removeSubscriptions(element: HTMLElement) {
        if(!!element.attributes) {
            for(let i = 0; i < element.attributes.length; i++) {
                if(element.attributes[i].name.indexOf('on') === 0) {
                    element.removeAttribute(element.attributes[i].name);
                }
            }
        }
        if(!!element.childNodes) {
            for(let i = 0; i < element.childNodes.length; i++) {
                removeSubscriptions(element.childNodes[i] as HTMLElement);
            }
        }
    }

    function fixHtmlXSS(value: string): string {
        let frame = document.getElementById('my_frame') as HTMLIFrameElement;
        if(!frame) {
            frame = document.createElement('iframe');
            frame.style.display = 'none';
            (frame as any)['sandbox'] = 'allow-same-origin';
            frame.id = 'my_frame';
            document.body.appendChild(frame);
        }
        frame.contentWindow.document.open();
        frame.contentWindow.document.write(
            `<head>
            </head>
            <body>
                ${value}
            </body>`
        );
        frame.contentWindow.document.close();
        removeSubscriptions(frame.contentWindow.document.body);
        let scripts = frame.contentWindow.document.body.getElementsByTagName('script');
        for(let i = 0; i < scripts.length; i++) {
            scripts[i].remove();
        }
        return frame.contentWindow.document.body.innerHTML;
    }

```

Maybe this solution can help someone.

**Comment by Saduff** [comment]
This can be trivially bypassed, see this PoC:
https://codepen.io/Saduff/pen/xxjMaQy

I've got a much better solution for everyone: stop including user-generated HTML on your pages! Or if you really must, use a proven solution such as [DOMPurify](https://github.com/cure53/DOMPurify) instead of reinventing the wheel. However, to use Quill safely, you don't need DOMPurify or anything like that.

---

> Yes that’s right, two alert dialog boxes.

And for those addicted to alert boxes, here's another PoC without any alerts that once again proves Quill is not vulnerable:
https://codepen.io/Saduff/pen/poVGOYr

And before you can say "but of course there are no alerts, because it's in a sandboxed iframe", let me stop you right there. That's not the point at all. The point is this:
![image](https://user-images.githubusercontent.com/11700507/195720523-77a496b4-0b46-4b34-9a73-0be26b66fd49.png)

As you can see, Quill created DOM contains no JavaScript that was included in the iframe `srcdoc`! Quill added the `<img>` tag without the `onerror` event handler, stripped the `<script>` tag, leaving only the text content and replaced the `<a>` tag `href` attribute with `about:blank`.

**Comment by BarkovskiyMaxim** [comment]
Thank you for a example, but its works good in quill, because quill can sanitize this code. Any way, solution with `iframe` works fine.

**Comment by Saduff** [comment]
> Quill try execute this events, but can't do this in iframe and sanitize template in next step.

Quill is not executing these events, the browser is. Quill has not been initialized at that point.

**Comment by tjad** [comment]
Perhaps this could be cleared up with better documentation.

If you use Quill instance's `getText()` method, you'll receive a WYSIWYG version of the contents.
If you want completely santised text, you can use the Quill instance's `root` property to get the innerHTML, which is fully sanitised.

**Comment by tjad** [comment]
Perhaps this could be cleared up with better documentation.

If you use Quill instance's `getText()` method, you'll receive a WYSIWYG version of the contents.
If you want completely santised text, you can use the Quill instance's `root` property to get the innerHTML, which is fully sanitised.

@snoopysecurity

**Comment by twocs** [comment]
> Perhaps this could be cleared up with better documentation.
> 
> If you use Quill instance's `getText()` method, you'll receive a WYSIWYG version of the contents. If you want completely santised text, you can use the Quill instance's `root` property to get the innerHTML, which is fully sanitised.
> 
> @snoopysecurity

I don't believe that documentation is sufficient. You could similarly restrict your page to no-inline or require nonces for inline scripts. It is not sufficient for there to be an option or other way to sanitise the input, because a module like quill is typically a third-party import that the typical developer is not an expert. It was only after several months that we realised that the library we use uses quill. To alleviate the security issue, the default must be to sanitise to prevent XSS; it is not sufficient that there is an option or other way to sanitise, that option must be the standard way to use the library.
 
The typical developer is using a library and will consider it to be reasonably secure. For example, OWASP suggests that:
 > Start with using your framework’s default output encoding protection when you wish to display data as the user typed it in. Automatic encoding and escaping functions are built into most frameworks. https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html

A similar library was marked as vulnerable by Snyk for a similar issue (now resolved). This article 
> Snyk flagged the "marked" package as vulnerable because although it had XSS protections, it was turned off by default. https://snyk.io/blog/marked-xss-vulnerability/

**Comment by tjad** [comment]
Simply put, santisation (XSS) is not the responsibility of an editor (unless the editor was responsible for a presentation layer).

Sanitisation is responsibility of the presentation layer.

In Quill's case, its presentation layer does not permit any XSS hacks (nor would it - as the editor is not the person trying to inject a vulnerability into their own browser).

**Comment by Saduff** [comment]
> To alleviate the security issue, the default must be to sanitise to prevent XSS; it is not sufficient that there is an option or other way to sanitise, that option must be the standard way to use the library.

That _is_ the default behavior. What other way are you talking about?

> For these reasons, my opinion is that the way to fix this issue is for Quill to filter the input as the default

Which it does, as my now 3 PoCs have shown, the last one with the `<template>` being the best way to demonstrate it. To be clear, these are just proof of concepts that are absolutely _not_ the way to use Quill. Their only purpose was to prove a point.

**Comment by Saduff** [comment]
> Lets assume I have sanitized the inputs coming in from quill, how does that remove the dependency bot alert for this warning? do I just ignore it?

Yes, ignore.

> I need to show that there are no vulnerabilities in current packages I am using. Or is that what is up to debate that the depencybot should not be flagging this issue?

Yes, it's a false positive.

> What is the proper way to test that input is getting sanitized

This is the best way I found:
https://codepen.io/Saduff/pen/rNvbWzv

You don't need DOMPurify or anything like that for Quill. However, depending on what else you do in your application, you may need it or something similar, but that depends on the specifics of the application and is out of scope of this discussion.

**Comment by adamskwersky** [comment]
Question about this. If I do not allow direct HTML editing, and the input is always stored and retrieved from the server as a Quill Delta, does this in effect prevent the XSS?

**Comment by Saduff** [comment]
> Question about this. If I do not allow direct HTML editing, and the input is always stored and retrieved from the server as a Quill Delta, does this in effect prevent the XSS?

Yes, and that is how it should be done.

**Comment by twocs** [comment]
> > To alleviate the security issue, the default must be to sanitise to prevent XSS; it is not sufficient that there is an option or other way to sanitise, that option must be the standard way to use the library.
> 
> That _is_ the default behavior. What other way are you talking about?

Quill does not sanitise inline javascript, such as onerror, by default.

**Comment by Saduff** [comment]
> Quill does not sanitise inline javascript, such as onerror, by default. 

Can you provide an example of that?

**Comment by tjad** [comment]
@Saduff / @twocs This is true/confirmed.

On the 1 hand, I agree, when placing raw text into the editor which is unsanitised, it will result in the javascript executing.
However, this is not caused by the editor not sanitising html. The method of input is explictly through route of raw HTML, hence the user should be cognisant to ensure this input is sanitised.

**Comment by tjad** [comment]
what I mean by "not caused by the editor not sanitising html" , is that the editor is not outputting unsafe HTML.

**Comment by tjad** [comment]
i.e if you take any element and place raw HTML within its `innerHTML`, XSS will happen if the user is not sanitising the html. The explicit API method to do that is `Quill.clipboard.dangerouslyPasteHTML()` , but again, this is explicitly pasting in potentially dangerous HTML (which is left as responsibility on the user to ensure it is sanitised -  not the editor).

I would only consider this an XSS vulnerabiity if the output of the WYSIWYG editor is outputting non-safe HTML. If the user is putting in raw HTML that is unsafe through a mechanism purposefully not cleaning the input, you can't place that under responsibility of the editor not sanitising text. (Because the intent was to be raw).

**Comment by Saduff** [comment]
> I would only consider this an XSS vulnerabiity if the output of the WYSIWYG editor is outputting non-safe HTML.

Yes, exactly. Which is what I have tried to show with my PoCs, i.e. the editor is outputting safe HTML.

> The explicit API method to do that is Quill.clipboard.dangerouslyPasteHTML()

I was not aware of this method, but it doesn't change anything. Where is this documented? This seems like an internal undocumented method not designed to be used by users. It's not mentioned at https://quilljs.com/docs/api/

**Comment by tjad** [comment]
@twocs Moreover, if you utilise the Quill Delta API as it intends you to use. It will sanitise the input.

e.g:

```
        quill.setContents([
          { insert: {image: 'https://quilljs.com/images/cloud.png'}, attributes:{onerror:'console.log(1337)', width:45} },
        ]);
      
```

**Comment by twocs** [comment]
There are many variations on XSS exploits. The risk I'm seeing here is that quill allows malicious code to run in the trusted environment, not that the final output of Quill contains the malicious code. In other words, "A vulnerability in the HTML editor of Slab Quill allows an attacker to execute arbitrary JavaScript by storing an XSS payload (a crafted onloadstart attribute of an IMG element) in a text field". The issue is not that after Quill is finished its business, the output of the editor still contains the XSS payload.

It's possible to use setContents with properly crafted input to pop up an alert: https://codepen.io/twocs/pen/NWzWrer, so the exploit simply needs quill to run quill.clipboard.convert on the malicious XSS payload. 

```
var potentiallyMaliciousCode = '<img src=ffs onerror="alert(12345)">'
var contents = quill.clipboard.convert(potentiallyMaliciousCode);
```

Doesn't even require `dangerouslyPasteHTML`. Using `this.clipboard.convert` is literally the same pattern that is used in the Quill Playground, so I'd assume that's how a developer who is familiar with Quill will use it. Indeed, the typical pattern is that when Quill loads up, it runs the `this.clipboard.convert(html)` on whatever is in the #editor on the html page.

**Comment by Saduff** [comment]
Thanks for that example. I can see why that would be a concern. However, I don't think this is an issue in this case.

I found that `quill.clipboard.dangerouslyPasteHTML` is actually documented here: https://quilljs.com/docs/modules/clipboard/

However, `quill.clipboard.convert` is not. So that is an internal undocumented method and can be considered equivalent to `dangerouslyPasteHTML` which is equivalent to setting the `innerHTML` of the editor.

> Using `this.clipboard.convert` is literally the same pattern that is used in the Quill constructor, so I'd assume that's how a developer who is familiar with Quill will use it.

Based on what I wrote above, I don't think so. A developer would use documented methods, not look through the source code of the library and use internal undocumented methods. In my pentest case, the developers did not use a Quill method at all and simply set the `innerHTML` of the editor using jQuery.

> Indeed, the typical pattern is that when Quill loads up, it runs the `this.clipboard.convert(html)` on whatever is in the #editor on the html page.

This is also not an issue. The reason why is already in your sentence: _whatever is in the #editor on the html page._

What would be a vulnerability is if you can paste an XSS payload in the editor and it would execute. However, I have not seen evidence of that.

**Comment by adamskwersky** [comment]
> It's possible to use setContents with properly crafted input to pop up an alert: https://codepen.io/twocs/pen/gOKOLGe, so the

I looked at your codepen but it wasn't using setContents. I changed it to setContents and the alert no longer pops up. Also if I use setText, the alert doesn't pop up.
Are you able to provide an example with setContents?

**Comment by twocs** [comment]
> > It's possible to use setContents with properly crafted input to pop up an alert: https://codepen.io/twocs/pen/gOKOLGe, so the
> 
> I looked at your codepen but it wasn't using setContents. I changed it to setContents and the alert no longer pops up. Also if I use setText, the alert doesn't pop up.
> Are you able to provide an example with setContents?
> 

Sorry, that's not how setContents would work.
`quill.clipboard.convert` is used to convert arbitrary HTML to a delta, which you would then feed into one of the set functions. I stripped the example down to its minimal problem to avoid distraction.

Many projects store the output from Quill as html. They similarly load that output back into the editor for editing. This indicates that if the hacker can get the malicious payload into the database, then the process of converting the HTML to a delta to display in the editor, Quill will run the code.

**Comment by adamskwersky** [comment]
> I feel that Quill could convert HTML more safely. It should be possible to convert HTML into elements and attributes without rendering.

Thanks for the explanation. Do you know if it is possible to copy from another web page and paste into Quill and have any of the JavaScript or onclick code included in the paste?

**Comment by twocs** [comment]
The problem seems to be caused by innerHTML here:
https://github.com/quilljs/quill/blob/1.3.7/modules/clipboard.js#L77

Effectively this is being run: 
`document.createElement('div').innerHTML = potentiallyMaliciousCode;`

[Owasp article on mitigating XSS uses innerHTML as its example.](https://cheatsheetseries.owasp.org/cheatsheets/DOM_based_XSS_Prevention_Cheat_Sheet.html)

The new version of quill.clipboard.convert (develop branch) no longer uses innerHTML, so would appear they will have addressed the issue in the upcoming major release version 2.

**Comment by Saduff** [comment]
> The problem seems to be caused by innerHTML here:
https://github.com/quilljs/quill/blob/1.3.7/modules/clipboard.js#L77

Yes, I found this gets called when I was creating my first PoC. However, I did not think of it as problematic, because it's only called in the constructor here: https://github.com/quilljs/quill/blob/1.3.7/core/quill.js#L100. The constructor doesn't take HTML input.

`quill.clipboard.convert` is not supposed to be called by users of the library. If that's followed, there is no vulnerability. However, I agree it would be better to avoid using `innerHTML` there. Quill shouldn't even expose this function to users.

**Comment by twocs** [comment]
> Yes, I found this gets called when I was creating my first PoC. However, I did not think of it as problematic, because it's only called in the constructor here: https://github.com/quilljs/quill/blob/1.3.7/core/quill.js#L100. The constructor doesn't take HTML input.

Whether or not you think it's problematic, if it can be even a small part of a complex hack, then it's a vulnerability.

**Comment by Saduff** [comment]
> Whether or not you think it's problematic, if it can be even a small part of a complex hack, then it's a vulnerability.

It can only be exploited if the developers (users of Quill) call `quill.clipboard.convert` explicitly with user input, which shouldn't be done as it's an undocumented internal function. But it's good to see they're not using `innerHTML` in the next version of Quill.

**Comment by tjad** [comment]
> > I feel that Quill could convert HTML more safely. It should be possible to convert HTML into elements and attributes without rendering.
> 
> Thanks for the explanation. Do you know if it is possible to copy from another web page and paste into Quill and have any of the JavaScript or onclick code included in the paste? I want to support copying and pasting from web pages (not the JS), just the static content, and my back-end stores the content as the JSON stringify of the Delta.

As you are storing the Delta and using `getContent()`, `setContent(), your usage is 100% safe.

 For 1.3.7, theoretically, if browsers copied the JS of attributes too, you'd be vulnerable to XSS if the alternative site that you select-copied from had XSS in it. However this is not a practical use-case for 2 reasons

1.  As far as I'm aware browsers do not copy such JS content, ,and this is only an exploit during pasting (hopefully you are not trying to exploit yourself?). 

2.  `getContent()` is still going to give you a sanitised Delta, and when inputting such Delta with `setContent()`, it will  be resanitised.

**Comment by tjad** [comment]
> The problem seems to be caused by innerHTML here: https://github.com/quilljs/quill/blob/1.3.7/modules/clipboard.js#L77
> 
> Effectively, when converting potentially malicious html to a delta, this is being run: `let potentiallyMaliciousCode = "<img src=q onerror=alert(1)>"; document.createElement('div').innerHTML = potentiallyMaliciousCode;`
> 
> [Owasp article on mitigating XSS uses innerHTML as its example.](https://cheatsheetseries.owasp.org/cheatsheets/DOM_based_XSS_Prevention_Cheat_Sheet.html)
> 
> The new version of quill.clipboard.convert (develop branch) no longer uses innerHTML, so would appear they will have addressed the issue in the upcoming major release version 2.  This seems to be the right way to allow conversion of pasted html into a delta without rendering it. Fixed code:   `const doc = new DOMParser().parseFromString(html, 'text/html');` https://github.com/quilljs/quill/blob/develop/modules/clipboard.ts#L117

Found this interesting, thanks.
And yes, agree with @Saduff. Use the tools correctly(as intended as per documentation), otherwise it is simply a 10T error.

**Comment by tjad** [comment]
> > Yes, I found this gets called when I was creating my first PoC. However, I did not think of it as problematic, because it's only called in the constructor here: https://github.com/quilljs/quill/blob/1.3.7/core/quill.js#L100. The constructor doesn't take HTML input.
> 
> Whether or not you think it's problematic, if it can be even a small part of a complex hack, then it's a vulnerability.

@twocs  Not quite, a vulnerability is an exposed attack surface. The surface here is not exposed in any practical sense. Unless a user exposes that surface through bad usage, how is it a vulnerability.  Or otherwise every package is exploitable - through bad usage.

**Comment by tjad** [comment]
If this was such an exploit, why would so many large vendors back Quill ? 
Grammarly, LinkedIn, Microsoft, Salesforce, Slack ....

**Comment by pauldraper** [comment]
@tjad unfortunately, I don't think you're going to be able to countermand the CVE in this thread.

**Comment by tjad** [comment]
@pauldraper probably not, but I'm certain we have established a lot more clarity for any readers that this CVE is a false positive.

**Comment by burninatorsec2** [comment]
> If this was such an exploit, why would so many large vendors back Quill ? Grammarly, LinkedIn, Microsoft, Salesforce, Slack ....

Big vendors do have security issues, they have them all the time. In the original blog post, this actually WAS reported to LinkedIn, who responded appropriately.

**Comment by burninatorsec2** [comment]
At this point, I've completed the due diligence for demonstrating how the issue works and reporting it. For now I'll recommend that clients either write validation around it as best they can, or use a differently-designed wysiwyg editor, unless a patch or rewrite comes out. Thanks for the discussion everyone.

**Comment by Saduff** [comment]
> At this point, I've completed the due diligence for demonstrating how the issue works and reporting it. For now I'll recommend that clients either write validation around it as best they can, or use a differently-designed wysiwyg editor, unless a patch or rewrite comes out. Thanks for the discussion everyone.

No, you haven't demonstrated the issue, at all. You haven't produced a single PoC, not even in the original blog post.

**Comment by Tofandel** [comment]
For a demonstration of the issue:
https://codesandbox.io/s/cool-scooby-77pb5e?file=/src/App.vue

So basically the issue is that, the content going through quill is not sanitized, so if you are just sanitizing the output to a web page as you should always do and is not a CVE, but are using this same content as is in the quill editor without sanitization (which is quite possible) then you are affected by this CVE

But it should be a low enough CVE because it can be mitigated quite easily by using the same sanitization for the input of quill as for what you would output as raw html

**Comment by Saduff** [comment]
@Tofandel, that does not demonstrate a vulnerability in Quill. However, it does demonstrate a very real vulnerability in https://github.com/vueup/vue-quill.

From the [v-model:content documentation](https://vueup.github.io/vue-quill/api/#v-model-content) of vue-quill (emphasis mine):

> Two-way binding editor content, can be Delta object, plain text, or **html string**

Quill does not support setting editor content via a HTML string. So in order to add that support, vue-quill has introduced a vulnerability at https://github.com/vueup/vue-quill/blob/dc589e1f1a1d1f76e717abf7f031dc43dcfcadeb/packages/vue-quill/src/components/QuillEditor.ts#L323-L325

**Comment by Tofandel** [comment]
@Saduff  Vanilla quill has the same vulnerability
https://codepen.io/Tofandel/pen/zYJYmOL

**Comment by Saduff** [comment]
I agree that there is a vulnerability if you call `clipboard.convert`, `pasteHTML` or `dangerouslyPasteHTML`. What I don't understand is, why would anyone think that's an acceptable thing to do (i.e. I have not found any documentation about these functions aside from `dangerouslyPasteHTML` which has 'dangerous' in its name and a disclaimer about XSS in the docs). Frameworks like React have similar functions for introducing XSS vulnerabilities.

> both pasteHTML and clipboard.convert which the doc implicitely advertised as the safe way to do that as opposed to dangerouslyPasteHTML

Can you point to where you found that in the docs?

**Comment by Tofandel** [comment]
> why would anyone think that's an acceptable thing to do
Because it's the only provided way to work with HTML **both ways** in quill, which is y'know kinda what you need to output on the frontend, not everybody can work with delta as an input

But on what we agree on is that it's not acceptable to not escape on the server the content you're gonna send to these functions, but I can think of a few cases where that might not be possible, like collaborative editing over websocket where that content doesn't directly go through a server (though using delta is probably a better format there) 

I meant implicitely as not documented, but given one function is named `dangerouslyPasteHTML` one can _mistakenly_ assume that `pasteHTML` is the safe version, though it's just an alias

Non documented code necessary for basic functionnality has a way to write itself in forums, blog posts, stackoverflow etc, and in this case it's exactly what happened as that documentation has been made up using either `.innerHTML`, `pasteHTML`, `clipboard.dangerouslyPasteHTML` or `clipboard.convert`

**Comment by Saduff** [comment]
> Because it's the only provided way to work with HTML both ways in quill, which is y'know kinda what you need to output on the frontend, not everybody can work with delta as an input, especially if you've migrated your content from some older html

That's a good point. I'm not as familiar with the use cases as I'm not a Quill user myself. However, this seems to me more like a misuse of Quill for something it was not designed for, so another rich text editor might be a better choice for such a use case (or perhaps a newer version of Quill, if such a version exists).

> But on what we agree on is that it's not acceptable to not sanitize server-side the content you're gonna send to these functions, but I can think of a few cases where that might not be possible, like collaborative editing over websocket where that content doesn't directly go through a server (though using delta is probably a better format there)

Contrary to popular belief, it is perfectly acceptable to sanitize client-side. Client-side is actually preferable for client-side libraries. Server-side is more for when HTML is rendered on the server, which is not so common these days, as frontends are commonly made using a framework such as Vue, Angular or React.

> Non documented code necessary for basic functionnality has a way to write itself in forums, blog posts, stackoverflow etc

I agree that one would expect a documented API for inputting HTML to a rich text editor, but instead of abusing one that doesn't, I'd look elsewhere for one that does. But that's just me, there may be good reasons for sticking with Quill.

**Comment by twocs** [comment]
> I agree that there is a vulnerability if you call clipboard.convert, pasteHTML or dangerouslyPasteHTML... ~ @Saduff 

I mean, we are literally rehashing the conversation that's been going on since 2021. The `clipboard.convert` does not imply you're pasting into the field. It's used in the core Quill js for the setup.
https://github.com/quilljs/quill/blob/d2f689fb4744cdada96c632a8bccf6d476932d7b/core/quill.ts#:~:text=%3D%20this.-,clipboard.convert,-(%7B

The new version of Quill works and does not have the problem, but it seems like it's never going to be , despite in 2021 the docs heading being updated to say "Note: This branch and README covers the upcoming 2.0 release. View [1.x docs here](https://github.com/quilljs/quill/tree/1.3.6)." Quill core in 1.x is using the clipboard.convert function in the setup.

**Comment by Saduff** [comment]
> It's used in the core Quill js for the setup.

Yes, that is known, but it's irrelevant. Like I said last year:

> Yes, I found this gets called when I was creating my first PoC. However, I did not think of it as problematic, because it's only called in the constructor here: https://github.com/quilljs/quill/blob/1.3.7/core/quill.js#L100. The constructor doesn't take HTML input.

Look at the source of the HTML which is later passed as input to the `convert` function:
https://github.com/quilljs/quill/blob/d2f689fb4744cdada96c632a8bccf6d476932d7b/core/quill.ts#L160

It's the container's `innerHTML`. What this means is:

* If all you do is switch to Quill v2 where the issue is fixed, your app is still vulnerable
* If you fix the issue in your app, the Quill version no longer matters

In short, the issue is not within Quill, the source of the vulnerability is outside of Quill. If you don't fix the vulnerability in your app, switching to Quill v2 will only prevent the second execution of the XSS, which doesn't matter from an attacker's perspective.

It would be a different case if the Quill constructor took a HTML string as input. In that case, Quill will be the only one who will execute the XSS. So, like I've said before, there is no reason for Quill to set the `innerHTML` of the container in the `convert` function, so v2 is an improvement.

**Comment by twocs** [comment]
The "container" html in v2 is the same as the container in v1. However, the clipboard.convert function no longer uses the DOM to get the content, instead using DOMParser. 
https://github.com/quilljs/quill/blob/d2f689fb4744cdada96c632a8bccf6d476932d7b/modules/clipboard.ts#L93

DOMParser alone will not keep you safe, because the xss still exists in the html, but from there, Quill v2 then converts the html into a Quill Delta before it is inserted into the DOM. From my experimentation, it doesn't perform the XSS replay that Quill v1 is prone to doing, so you will only see one alert message, not two.

**Comment by Saduff** [comment]
> Non documented code necessary for basic functionnality has a way to write itself in forums, blog posts, stackoverflow etc, and in this case it's exactly what happened as that documentation has been made up using either .innerHTML, pasteHTML, clipboard.dangerouslyPasteHTML or clipboard.convert

This is very true. Just recently during a pentest, I found the developers doing the following in their Angular application:
```javascript
quill.setContents(quill.clipboard.convert(this.value));
```

Which as we know is vulnerable, but should be fixed in Quill v2.

**Comment by tysonclugg** [comment]
This issue appears to be fixed in https://github.com/slab/quill/pull/2226/commits/16b6a6cd842056f06c5ac2fc49ff9ac39dd17409 (via PR https://github.com/slab/quill/pull/2226) as part of [v2.0.0](https://github.com/slab/quill/releases/tag/v2.0.0).

I'm no expert, but does this mean a the original CVE should be re-issued with a new number?

I'm suggesting the security advisory be re-issued because:
1. You've fixed the issue, well done!
2. [It's good practice to issue security advisories against your own software](https://github.blog/security/vulnerability-research/removing-the-stigma-of-a-cve/).
3. There doesn't seem to be any way to retract the DISPUTED status of a CVE.
4. Users of tools like Synk may have [ignored this issue](https://docs.snyk.io/snyk-cli/commands/ignore) based on the DISPUTED status of the original CVE, their security would be best served by having a new advisory with a new number that isn't in their ignore list, so they receive notification to upgrade to a patched version.

**Comment by ChristianPilot** [comment]
Using the raw editor at Quill here in the quickstart: https://quilljs.com/docs/quickstart

Also has an XSS vulnerability?

You can pase this in:


<link href="https://cdn.jsdelivr.net/npm/quill@2.0.3/dist/quill.snow.css" rel="stylesheet" />


<div id="editor">
  <p><script>alert("Hello World!")</script></p>
  <p>Some initial <strong>bold</strong> text</p>
  <p><br /></p>
</div>


<script src="https://cdn.jsdelivr.net/npm/quill@2.0.3/dist/quill.js"></script>


<script>
  const quill = new Quill('#editor', {
    theme: 'snow'
  });
</script>

And it will execute the alert?
